### ISSUE TITLE
illegalmoveが具体的にどのようなものか明記されていない

### ISSUE THREAD

**Issue by stnk20** [TukamotoRyuzo/upi-protocol#2]
illegalmoveについて、具体的にどのようなものかが明記されていません。

明らかなものとして下記がありますが、
* 空中への配置
* すでにぷよがある位置への配置
* 軸ぷよと子ぷよが1列以上離れての配置

次のような場合をはどうでしょうか？
* 14段目への軸ぷよの配置
* 落下中の軸ぷよよりも高い位置への配置
* 13段目まで積み上がった列を越えての配置
* 12段目まで積み上がった列を越えての配置

実際のゲームとの一致を重視していないことは理解しますが、実際のゲームとの違いは明記しておく必要があると思われます。

**Comment by TukamotoRyuzo** [comment]
そうですね。まずillegalではないmoveというのが何か、というのを定義し、そこに含まれないmoveはすべてillegalである、と説明するのがよさそうですね。
考えて修正してみます。

**Comment by stnk20** [comment]
落下中のぷよ位置に依存してillegalmoveが決まるようにするのであれば、pfenに現在のぷよ位置を記載する必要が出てくると思います。
割り切って、落下中のぷよという概念をなくすことも考えられます。そうすれば移動可能な範囲＝操作可能な範囲となります。ただその場合でも、距離やちぎりによる時間差などは反映すべきだと思います（これをしないとほぼターン制になってしまいます）。

**Comment by TukamotoRyuzo** [comment]
> 落下中のぷよ位置に依存してillegalmoveが決まるようにするのであれば、pfenに現在のぷよ位置を記載する必要が出てくると思います。

はい、そうですね。

> 割り切って、落下中のぷよという概念をなくすことも考えられます。そうすれば移動可能な範囲＝操作可能な範囲となります。

すみません、これがちょっとよくわかりません。移動可能な範囲は落下中のぷよの位置によって決まると思うのですが、その概念をなくすというのはどういう意味なんでしょう？

**Comment by stnk20** [comment]
すいません、もう少し説明します。
まず、移動可能な範囲→**3lから**移動可能な範囲と読み替えてください。
実環境のほうは、高速落下中に相手の手を見て手を変更するなどできます（現状のプロトコルでは相手の手を見るためには自由落下しないといけないと理解しています）。
現状のプロトコルでは操作可能になる時点までにbestmoveを送る構成になっていましたので、上記の振る舞いに対応するには煩雑すぎるかなと感じたため、いっそ落下中のぷよという考えをなくしてしまえばシンプルかなと考え、あくまで一つの手段のつもりで記載しました。
